crypto/dh/dh_pmeth.c

```diff
@@ -114,11 +114,22 @@ static int pkey_dh_init(EVP_PKEY_CTX *ctx)
     return 1;
 }
 
+static void pkey_dh_cleanup(EVP_PKEY_CTX *ctx)
+{
+    DH_PKEY_CTX *dctx = ctx->data;
+    if (dctx != NULL) {
+        OPENSSL_free(dctx->kdf_ukm);
+        ASN1_OBJECT_free(dctx->kdf_oid);
+        OPENSSL_free(dctx);
+        ctx->data = NULL;
+    }
+}
+
 static int pkey_dh_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
 {
     DH_PKEY_CTX *dctx, *sctx;
     if (!pkey_dh_init(dst))
-        return 0;
+        goto err;
```

Comment:
Invoking pkey_dh_cleanup(dst) here on error, expects dst->data to be set to NULL by the callee 
Otherwise, here it may just return 0, as nothing to be freed or make pkey_dh_init(dst) initialises dst->data to NULL on error.